lyncs_io/hdf5.py

```diff
@@ -8,72 +8,109 @@
+    if comm is not None:
+        _, subsizes, starts = Decomposition(comm=comm).decompose(dts.shape)
+        slc = tuple(slice(start, start + size) for start, size in zip(starts, subsizes))
+    else:
+        slc = tuple(slice(size) for size in dts.shape)
 
-    return from_array(dts[:], attrs)
+    return attrs, from_array(dts[slc], attrs)
 
 
-def _load(h5f, depth=1, header_only=False, **kwargs):
+def _load(h5f, depth=1, header_only=False, all_data=False, **kwargs):
     if isinstance(h5f, Group):
         return {
-            key: _load(val, depth=depth - 1) if depth > 0 else None
+            key: _load(val, depth=depth - 1, all_data=all_data, **kwargs)
+            if all or depth > 0
+            else None
             for key, val in h5f.items()
         }
 
     if isinstance(h5f, Dataset):
-        header = _load_dataset(h5f, header_only=True, **kwargs)
+        header, data = _load_dataset(
+            h5f, header_only=header_only or (not all_data), **kwargs
+        )
+        if header_only:
+            return header
+        return Data(header, data)
+
+    raise TypeError(f"Unsupported {type(h5f)}")
+
+
+def _load_dispatch(h5f, key, loader, header_only=False, **kwargs):
+
+    if key:
+        h5f = h5f[key]
+
+    if isinstance(h5f, Dataset):
+        header, data = _load_dataset(h5f, **kwargs)
         if header_only:
             return header
-        return Data(header)
+
+        return data
+
+    if isinstance(h5f, Group):
+        return Archive(_load(h5f, **kwargs), loader=loader, path=key)
 
     raise TypeError(f"Unsupported {type(h5f)}")
 
 
-def load(filename, key=None, **kwargs):
+def load(filename, key=None, chunks=None, comm=None, **kwargs):
     "Load function for HDF5"
-    filename, key = split_filename(filename, key)
+    if comm is not None and chunks is not None:
+        raise ValueError("chunks and comm parameters cannot be both set")
 
+    filename, key = split_filename(filename, key)
+    # Append comm and chunks in kwargs
+    kwargs = {"comm": comm, **kwargs}
     loader = Loader(load, filename, kwargs=kwargs)
 
-    with File(filename, "r") as h5f:
-        if key:
-            h5f = h5f[key]
+    if chunks is not None:
+        raise NotImplementedError("DaskIO for HDF5 load not implemented yet.")
 
-        if isinstance(h5f, Dataset):
-            return _load_dataset(h5f, **kwargs)
+    if comm is not None:
+        check_comm(comm)
 
-        if isinstance(h5f, Group):
-            return Archive(
-                _load(h5f, **kwargs),
-                loader=loader,
-                path=key,
-            )
+        if comm.size > 1:
+            with File(filename, "r", driver="mpio", comm=comm) as h5f:
+                return _load_dispatch(h5f, key, loader, **kwargs)
 
-        raise TypeError(f"Unsupported {type(h5f)}")
+    with File(filename, "r") as h5f:
+        return _load_dispatch(h5f, key, loader, **kwargs)
 
 
-def head(*args, **kwargs):
+def head(*args, comm=None, **kwargs):
```

Comment:
Does it make sense to have comm for `head`?
I don't see it and I would be in favor of removing it.

Reply:
No it doesn't really make sense to have it explicitly there. In case it is needed it is passed through kwargs anyway. Removed.